### PR TITLE
Feature/elastic search

### PR DESCRIPTION
**Pull request by DmitryTumanov**
Add logs of product to elasticsearch

### CODE REVIEW

File: OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/ClientsFabric/IClientsFabric.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using Nest;
+
+namespace OnlinerTask.BLL.Services.ElasticSearch.ProductLogger.ClientsFabric
+{
+    public interface IClientsFabric
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/ClientsFabric/IClientsFabric.cs]: Factory?

File: OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Threading.Tasks;
+using OnlinerTask.Data.SearchModels;
+
+namespace OnlinerTask.BLL.Services.ElasticSearch.ProductLogger
+{
+    public interface IProductLogger
+    {
+        Task LogAdding(int productId, ProductModel productModel);
+
+        Task RemoveLog(int productId);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs]: Is this method really need?

File: OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/Implementations/OnlinerProductLogger.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        private readonly ElasticClient elasticClient;
+
+        public OnlinerProductLogger(IClientsFabric clientsFabric, IConnectionFabric connectionFabric)
+        {
+            var settings = connectionFabric.CreateConnection(Configurations.ElasticSearchUrl, Configurations.ProductLogIndex);
+            elasticClient = clientsFabric.CreateClient(settings, Configurations.ProductLogIndex);
+        }
+
+        public async Task LogAdding(int productId, ProductModel productModel)
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/Implementations/OnlinerProductLogger.cs]: Names are not so good. Also think about common logging mechanism. On each model type you should not implement another one method.

File: OnlinerTask/OnlinerTask.BLL/Services/Products/Implementations/Manager.cs [hunk 4]
```diff
@@ -25,14 +28,16 @@ protected Manager(ISearchService searchService, IRepository repository)
             {
                 return;
             }
-            repository.CreateOnlinerProduct(result, name);
+            var productId = repository.CreateOnlinerProduct(result, name);
             AddNotify(result.FullName);
+            await productLogger.LogAdding(productId, result);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Products/Implementations/Manager.cs]: Logging should be deferred as much as possible. Try to rethink how logs should be stored.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

File: OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs (hunk 2 above)
Comment (DmitryTumanov) [reply OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs]: I think, it`s unnecessary to have log of product, which has been unchecked(and in this case we will always have logs of products, that interesting to users)

Comment (andrei-piatrou) [reply OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs]: So lets add another type of logs - product deletion. It is also useful for business to have this kind of info.

Comment (DmitryTumanov) [reply OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/IProductLogger.cs]: Okey

File: OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/ClientsFabric/IClientsFactory.cs [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace OnlinerTask.BLL.Services.ElasticSearch.ProductLogger.ClientsFabric
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ElasticSearch/ProductLogger/ClientsFabric/IClientsFactory.cs]: namespace?

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

**Review by andrei-piatrou — APPROVED** [review]

File: OnlinerTask/OnlinerTask.WEB/Controllers/PersonalController.cs [hunk 6]
```diff
@@ -6,10 +6,12 @@
 using OnlinerTask.BLL.Services.Products;
 using OnlinerTask.BLL.Services.TimeChange;
 using OnlinerTask.Data.Responses;
+using OnlinerTask.WEB.Filters;
 
 namespace OnlinerTask.WEB.Controllers
 {
     [Authorize]
+    [WebApiLogAction]
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.WEB/Controllers/PersonalController.cs]: it is easier to add this filter as global filter.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Pull request merged**